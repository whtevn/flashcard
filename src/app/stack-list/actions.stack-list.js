const Type = "Stack List"
export const SELECT = `${Type} SELECT`;
export const Select = ( payload ) => {
  return {
    type: SELECT,
    payload
  }
}

export const CREATE = `${Type} CREATE`;
export const Create = ( payload ) => {
  return {
    type: CREATE,
    payload
  }
}


