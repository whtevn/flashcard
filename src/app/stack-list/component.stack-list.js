import React from 'react'
import FCD_StackListElement from '../stack-list-element';

class FCD_StackList extends React.Component {
  render (){
    const stacks = this.props.stacks;
    const onSelect = this.props.onSelect;
    return (
      <div >
        { stacks.map((stack) => < FCD_StackListElement name={stack.get('name')} id={stack.get('key')} key={stack.get('key')} onSelect={ onSelect } cards={stack.get('cards')} /> ) }
      </div>
    )
  }
}
export default FCD_StackList;

