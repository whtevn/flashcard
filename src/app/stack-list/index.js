import FCD_StackList from './container.stack-list';
export default FCD_StackList;
export { Reducer, initialState } from './reducer.stack-list';
