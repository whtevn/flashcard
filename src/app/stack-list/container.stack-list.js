import { fromJS } from 'immutable';
import { connect } from 'react-redux';
import { default as Component } from './component.stack-list';
import * as Actions from './actions.stack-list';

// connect component to store
export function mapStateToProps(state, ownprops){
  state = state.stacks || state
  return {
    stacks: state.getIn(['data', 'stacks']),
    selected: state.getIn(['data', 'selected'])
  }
}

export function mapDispatchToProps(dispatch){
  return {
    onSelect: (selected, cards) => {
      dispatch(Actions.Select( fromJS({selected, cards}) ));
    }
  }
}

const FCD_StackList = connect(mapStateToProps, mapDispatchToProps)(Component);
export default FCD_StackList;

