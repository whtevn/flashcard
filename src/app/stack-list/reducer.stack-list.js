import { fromJS } from 'immutable';
import * as Actions from './actions.stack-list'

export const initialState = fromJS({
  data: {
    stacks: [],
    selected: undefined
  }
})

export const Reducer = (state = initialState, action) => {
  switch (action.type) {
    case Actions.SELECT:
      return state.setIn(['data', 'selected'], action.payload.get('selected'));
    case Actions.CREATE:
      const stacks = state.getIn(['data', 'stacks']);

      console.log(action.payload.toJS())
      return state.setIn(['data', 'stacks'], stacks.push(
        action.payload.get('stack').merge(fromJS({
          cards: [],
          key: stacks.size
        }))))
    default:
      return state;
  }
}
