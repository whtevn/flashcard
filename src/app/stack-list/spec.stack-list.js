import renderer from 'react-test-renderer';
import React from 'react';
import { Provider } from 'react-redux'
import { createStore } from 'redux'
import MuiThemeProvider from 'material-ui/styles/MuiThemeProvider';
import {default as Component} from './component.stack-list';
import * as Container from './container.stack-list';
import FCD_StackList, { initialState } from './';
import { List, Map, fromJS } from 'immutable';
import { Reducer } from './reducer.stack-list';
import * as Actions from './actions.stack-list';
const mockFn = jest.fn();

describe("The stack list", ()=>{
  let stackCards;
  let stackList;
  let store;
  beforeEach(()=>{
    stackCards = List()
                  .push(Map({question: "q 1", answer: "a 1"}))
                  .push(Map({question: "q 2", answer: "a 2"}))
                  .push(Map({question: "q 3", answer: "a 3"}))
                  .push(Map({question: "q 4", answer: "a 4"}))
                  .push(Map({question: "q 5", answer: "a 5"}))
    stackList = fromJS([
      {name: 'first stack' , key: 0, cards: stackCards },
      {name: 'second stack', key: 1, cards: stackCards },
      {name: 'third stack' , key: 2, cards: stackCards },
      {name: 'fourth stack', key: 3, cards: stackCards },
      {name: 'fifth stack' , key: 4, cards: stackCards }
    ])
    store = createStore(Reducer, {stacks: stackList});
  })

  describe("component", ()=>{
    test('Stack list renders correctly', () => {
      const tree = renderer.create(
            <Provider store={store}>
              <MuiThemeProvider>
                <Component stacks={stackList} onSelect={()=>{}} />
              </MuiThemeProvider>
            </Provider>
          ).toJSON();
      expect(tree).toMatchSnapshot();
    });
  })

  describe("container", ()=>{
    let state, props, selected;
    beforeEach(()=>{
      selected = 0;
    })
    describe('mapping state to props', ()=>{
      beforeEach(()=>{
        state = { stacks: fromJS({data: {stacks: stackList, selected}}) };
        props = Container.mapStateToProps(state)
      })
      it("should set the stacks", ()=>{
        expect(props.stacks).toBe(stackList)
      })

      it("should set the selected", ()=>{
        expect(props.selected).toBe(selected)
      })
    })

    describe('mapping dispatch to props', ()=>{
      beforeEach(()=>{
        props = Container.mapDispatchToProps(mockFn)
        props.onSelect(selected, stackCards)
      })
      afterEach(() => {
        mockFn.mockClear();
      })
      describe("when a stack is selected", ()=>{
        it("should dispatch the select action", ()=>{
          expect(mockFn.mock.calls).toMatchSnapshot()
        })
      })
    })
  })

  describe("reducer", ()=>{
    let state, action, next, selected;
    describe('StackList', ()=>{
      beforeEach(()=>{
        selected = 100;
        state = fromJS({data: {stacks: stackList, selected: 0}});
        action = { type: Actions.SELECT, payload: fromJS({ selected }) }
        next  = Reducer(state, action);
      })
      describe("::SELECT", ()=>{
        it("should read the selected id from the payload and set it in the state", ()=>{
          expect(next.getIn(['data', 'selected'])).toBe(selected);
        })
      })
    })
  })
})

