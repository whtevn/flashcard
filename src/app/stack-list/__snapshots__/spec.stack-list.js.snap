// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The stack list component Stack list renders correctly 1`] = `
<div>
  <div>
    <a
      href="#"
      onClick={[Function]}
    >
       
      first stack
       
    </a>
  </div>
  <div>
    <a
      href="#"
      onClick={[Function]}
    >
       
      second stack
       
    </a>
  </div>
  <div>
    <a
      href="#"
      onClick={[Function]}
    >
       
      third stack
       
    </a>
  </div>
  <div>
    <a
      href="#"
      onClick={[Function]}
    >
       
      fourth stack
       
    </a>
  </div>
  <div>
    <a
      href="#"
      onClick={[Function]}
    >
       
      fifth stack
       
    </a>
  </div>
</div>
`;

exports[`The stack list container mapping dispatch to props when a stack is selected should dispatch the select action 1`] = `
Array [
  Array [
    Object {
      "payload": Object {
        "cards": Array [
          Object {
            "answer": "a 1",
            "question": "q 1",
          },
          Object {
            "answer": "a 2",
            "question": "q 2",
          },
          Object {
            "answer": "a 3",
            "question": "q 3",
          },
          Object {
            "answer": "a 4",
            "question": "q 4",
          },
          Object {
            "answer": "a 5",
            "question": "q 5",
          },
        ],
        "selected": 0,
      },
      "type": "Stack List SELECT",
    },
  ],
]
`;
